components/server/src/jobs/runner.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { DisposableCollection } from "@gitpod/gitpod-protocol";
+import { repeat } from "@gitpod/gitpod-protocol/lib/util/repeat";
+import { inject, injectable } from "inversify";
+import { RedisMutex } from "../redis/mutex";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { ResourceLockedError } from "redlock";
+import { reportJobCompleted, reportJobStarted } from "../prometheus-metrics";
+import { DatabaseGarbageCollector } from "./database-gc";
+import { OTSGarbageCollector } from "./ots-gc";
+import { TokenGarbageCollector } from "./token-gc";
+import { WebhookEventGarbageCollector } from "./webhook-gc";
+import { WorkspaceGarbageCollector } from "./workspace-gc";
+
+export const Job = Symbol("Job");
+
+export interface Job {
+    name: string;
+    lockId: string[];
+    frequencyMs: number;
+    run: () => Promise<void>;
+}
+
+@injectable()
+export class JobRunner {
+    @inject(RedisMutex) protected mutex: RedisMutex;
+
+    @inject(DatabaseGarbageCollector) protected databaseGC: DatabaseGarbageCollector;
+    @inject(OTSGarbageCollector) protected otsGC: OTSGarbageCollector;
+    @inject(TokenGarbageCollector) protected tokenGC: TokenGarbageCollector;
+    @inject(WebhookEventGarbageCollector) protected webhookGC: WebhookEventGarbageCollector;
+    @inject(WorkspaceGarbageCollector) protected workspaceGC: WorkspaceGarbageCollector;
+
+    public start(): DisposableCollection {
+        const disposables = new DisposableCollection();
+
+        const jobs: Job[] = [this.databaseGC, this.otsGC, this.tokenGC, this.webhookGC, this.workspaceGC];
```

Comment:
We'd usually use multiinject for this kind of plug-in pattern. But since we are not building a framework listing the jobs explicitly here seems good and straightforward.

Reply:
I've started with that implementation, but felt it's too much magic. I chatted through it with Gero and he also thought being explicit here is easier/cleaner.

If it becomes too cumbersome to add it/we forget, we can always touch this up.